### PR TITLE
CIRC-245 & CIRC-264: Title Level Requests for Available Items

### PR DESCRIPTION
**Pull request by mtraneis**
## Purpose
This PR implements both CIRC-245 (Create  and retrieve a Title Level Request for an instance with one available item), CIRC-264 (Create  and retrieve a Title Level Request for an instance with multiple available items), and incorporates the code from CIRC-267 (Create endpoint for Title Level Requests), which has been approved.

## Approach
For CIRC-245 and CIRC-264, the goal is to make a request by an instanceID (title-level ID).  Using that InstanceID, FOLIO needs to figure out the copies (or items) of the instance and places a request for any of these items with the rules specified in CIRC-245 and CIRC-264.  Since we don't know which item will be available to make the request, according to various rules around the item's state and type, request type, material type, and so on, this implementation will defer to the existing code to figure this out. The responsibility of the new code is to create a list of possible item requests derived from the instance ID and sends them to the existing code to attempt to make a request. As soon as a request is made
the result is sent back to the client, and the same thing happens if there are any errors. 

Here's a high level algorithm:

1. Find Items by InstanceId
	a) Look up the holdings associated with the instance
	b) Determines the items associated with the holdings
	c) Fetch items from items-storage
2. Find AvailableItems:
	if (any AvailableItems)
		Find location for each Available Item
		Find Items that have location that have matching service pointID (in other words, find the items that are at the location that has a servicePoint that was specified in the instance request)
			These will be ranked highest, Then other Available items.
3. Create request objects for the sorted Available item. Each item will have 3 requests (at this moment), one for each request type.
4. Place requests for Available Items

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [x] Were there any schema changes?
  - [x] Did any of the interface versions change?
  - [x] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.http.HttpClient;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.RoutingContext;
+
+public class RequestByInstanceIdResource extends Resource {
+  private UserRepository userRepository;
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: Why are these declared as fields rather than variables within the method?

I don't know vert.x well, so I might be wrong about this. This class is only instantiated once. If there are interleaving HTTP requests being handled I imagine they could interfere with each other by changing this shared state.

File: descriptors/ModuleDescriptor-template.json [hunk 2]
```diff
@@ -683,6 +683,16 @@
...
+        },
+        {
+          "methods": [
+            "POST"
+          ],
+          "pathPattern": "/circulation/requests/instances",
+          "permissionsRequired": [
+            "circulation.requests.instances.item.post"
+          ],
+          "modulePermissions": []
```
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json]: The implementation appears to make requests to other endpoints. Should the relevant permissions for these be declared here?

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+    availableItems.thenCompose(r -> r.after( collectionResult -> findItemsWithMatchingServicePointId(
+                                                                  pickupServicePointId, collectionResult, locationRepository)))
+      .thenApply( r -> r.next( itemsFound -> instanceToItemRequests(requestByInstanceIdRequest, itemsFound)))
+      .thenCompose( r -> r.after( requests -> placeRequests(requests, clients)))
+      .thenApply(r -> r.map(RequestAndRelatedRecords::getRequest))
+      .thenApply(r -> r.map(new RequestRepresentation()::extendedRepresentation))
+      .thenApply(CreatedJsonResponseResult::from)
+      .thenAccept(result -> result.writeTo(routingContext.response()))
+      .exceptionally( err -> {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: Under what circumstances are exceptions expected to fall through to here?

Unless they relate to writing the HTTP response, could we send them on to the client?

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+      );
+
+    return requestFromRepresentationService.getRequestFrom(currentItemRequest)
+      .thenCompose(r -> r.after(createRequestService::createRequest))
+      .thenCompose(r -> {
+          if (r.succeeded()) {
+            return CompletableFuture.completedFuture(r);
+          } else {
+            log.debug("Failed to create request for " + currentItemRequest.getString("id"));
+            return placeRequest(itemRequests, startIndex +1, createRequestService);
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: My concern with this approach is that request creation might fail for a variety of reasons, and not because the request is invalid. 

In those circumstances, the system will move on to the next item. It might partially create multiple requests.

I likely didn't manage to raise this early enough in the process, in response to your thorough early design comments, and it is likely too late to change it. As it would likely require significant refactoring of the existing code to make this more amenable to doing it without potential side effects.

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 1 above)
Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: These are fields because they are being used elsewhere and I didn't want to pass them around. I don't know vertx much so if it's better to declare them within the method then I will do that.

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 3 above)
Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: I was trying to be safe and add a catch-all handler and send back a harmless message to the client.

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 4 above)
Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: I'm hoping that we will get a chance to implement the feature for real (not this pseudo implementation) and we will get to do what is right.  For this simple implementation, it might not be worth it to do significant refactoring. However, if you know of a case that illustrates this example, I'd love to see how this code reacts to it.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: Thinking of an example, is relative easy. Recreating it in the system less so.

If the update of the item status or loan fails, then I imagine this code would move on to the next item. That could result in each item getting a new request.

I agree that it likely requires a lot of refactoring. The validation of the request would need separating from the creation. Then only when the validation succeeded, actually create the request.

Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: (I reckon there is significant complexity around a mixed instance and item queue)

File: descriptors/ModuleDescriptor-template.json (hunk 2 above)
Comment (mreno-EBSCO) [reply descriptors/ModuleDescriptor-template.json]: Yes. I added this with my branch, but I didn't know what permissions would need to be added at that time. We'll need to figure this out so we can supply them here.

File: src/test/java/api/support/http/ResourceClient.java [hunk 5]
```diff
@@ -235,7 +235,7 @@ public Response attemptCreate(Builder builder)
     client.post(urlMaker.combine(""), builder.create(),
       ResponseHandler.any(createCompleted));
 
-    return createCompleted.get(5, TimeUnit.SECONDS);
+    return createCompleted.get(500, TimeUnit.SECONDS);
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/http/ResourceClient.java]: Why has this timeout been changed to such a large number?

File: src/test/java/api/support/builders/HoldingBuilder.java [hunk 6]
```diff
@@ -85,7 +93,18 @@ public HoldingBuilder withCallNumber(String callNumber) {
       this.instanceId,
       this.permanentLocationId,
       this.temporaryLocationId,
-      callNumber
+      callNumber,
+      this.holdingsItems
+    );
+  }
+
+  public HoldingBuilder withHoldingsItems(JsonArray holdingsItems) {
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/builders/HoldingBuilder.java]: This does not appear to be used. Is this a hangover from the confusion caused by the GraphQL metadata?

File: src/test/java/api/support/builders/LocationBuilder.java [hunk 7]
```diff
@@ -18,7 +18,7 @@
   private final UUID campusId;
   private final UUID libraryId;
   private final UUID primaryServicePointId;
-  private final Set<UUID> servingServicePointIds;
+  private final Set<UUID> servicePointIds;
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/builders/LocationBuilder.java]: Why was this renamed?

File: src/test/java/api/support/fixtures/LocationsFixture.java [hunk 8]
```diff
@@ -82,6 +83,19 @@ public IndividualResource thirdFloor()
       locationExamples.thirdFloor());
   }
 
+  public IndividualResource fourthFloor()
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/fixtures/LocationsFixture.java]: This example is never used, can it be removed?

File: src/test/java/api/support/fixtures/LocationsFixture.java [hunk 9]
```diff
@@ -106,6 +120,24 @@ public IndividualResource mezzanineDisplayCase()
...
+  */
+  public IndividualResource mainFloor()
+    throws InterruptedException,
+    MalformedURLException,
+    TimeoutException,
+    ExecutionException {
+
+    final LocationExamples locationExamples = getLocationMultiServicePointsExamples();
+
+    return locationRecordCreator.createIfAbsent(
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/fixtures/LocationsFixture.java]: This could fit on a single line

File: src/test/java/api/support/examples/LocationExamples.java [hunk 10]
```diff
@@ -11,19 +12,25 @@
...
+  private final UUID tertiaryServicePointId;
 
   public LocationExamples(
     UUID nottinghamUniversityId,
     UUID jubileeCampusId,
     UUID businessLibraryId,
     UUID djanoglyLibraryId,
-    UUID primaryServicePointId) {
+    UUID primaryServicePointId,
+    UUID secondaryServicePointId,
```
Comment (marcjohnson-kint) [inline src/test/java/api/support/examples/LocationExamples.java]: The domain does not have the concept of secondary or tertiary service points, only a single primary for a location and any other that serve that location.

Could the final two parameters be replaced by a variable argument?

File: src/test/java/api/requests/RequestByInstanceIdResourceTests.java [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+
+import api.support.APITests;
+import api.support.builders.LocationBuilder;
+import api.support.fixtures.ItemExamples;
+import io.vertx.core.json.JsonObject;
+
+public class RequestByInstanceIdResourceTests extends APITests {
+
+  @Test
+  public void canTransformInstanceToItemRequests(){
```
Comment (marcjohnson-kint) [inline src/test/java/api/requests/RequestByInstanceIdResourceTests.java]: I'm finding this test difficult to follow. Can you help me understand this test better.

What does this test do? It seems to create multiple requests, which is confusing to me,  thought an instance level request maps to a single item level request?

What do `i` and `j` represent? What does the loop do?

I find that branching (e.g. `if`) statements in tests tends to reduce readability and in some cases can suggest non-deterministic behaviour, which leads to brittle tests.

File: src/main/java/org/folio/circulation/support/http/client/OkapiHttpClient.java [hunk 12]
```diff
@@ -56,7 +56,7 @@ public void post(
 
     addStandardHeaders(request);
 
-    request.setTimeout(5000);
+    request.setTimeout(15000);
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/support/http/client/OkapiHttpClient.java]: Why has this timeout been increased by such a degree?

Comment (mtraneis) [reply src/main/java/org/folio/circulation/support/http/client/OkapiHttpClient.java]: I had to increase them to run/debug locally because my laptop is slow or when I need to pause for debug, and I accidentally checked these high timeouts in. I will revert these timeouts.

File: src/test/java/api/requests/RequestByInstanceIdResourceTests.java (hunk 11 above)
Comment (mtraneis) [reply src/test/java/api/requests/RequestByInstanceIdResourceTests.java]: This test is basically a unit test that checks the step where it creates _n_*_instance_ total number of item requests. Recall that  it does not know which type of request to make, so it creates all possible requests for all request types for that one item. This test loops through two items using the **j** counter and for each item it loops through all possible request types with the **i** counter.

Other tests in this file needed to make use of various fixtures so this test class has to extend APITests. We can create another class for true unit-level tests if needed.

File: src/test/java/api/support/examples/LocationExamples.java (hunk 10 above)
Comment (mtraneis) [reply src/test/java/api/support/examples/LocationExamples.java]: Yes, will change

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 1 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: My belief would be that state could be corrupted if multiple requests come in at the same time. Does my example make sense, that a second HTTP request will change the variables and hence affect the first?

I think it would be safer to use variables within the method. I'm aware that leads to noisier code.

File: src/test/java/api/requests/RequestByInstanceIdResourceTests.java (hunk 11 above)
Comment (marcjohnson-kint) [reply src/test/java/api/requests/RequestByInstanceIdResourceTests.java]: > This test is basically a unit test that checks the step where it creates n*instance total number of item requests.

Why does it create multiple requests in the same test?

> Recall that it does not know which type of request to make, so it creates all possible requests for all request types for that one item.

Within a test, where the policies and records are controlled, it should be possible to know what type of request is expected to be created.

I don't follow why that isn't possible here?

> Other tests in this file needed to make use of various fixtures so this test class has to extend APITests. We can create another class for true unit-level tests if needed.

I think splitting them out would be useful. I'm aware that due to the way the code is structured it is difficult to inject fakes to write more unit like tests. 

If we need a style of tests that doesn't use the API, but require infrastructure, then I think we should create that separately. It might be that is out of scope for this work (I'm saying this a lot at the moment, and the effect on the code base is noticeable).

Comment (mtraneis) [reply src/test/java/api/requests/RequestByInstanceIdResourceTests.java]: > Within a test, where the policies and records are controlled, it should be possible to know what type of request is expected to be created.

I don't follow why that isn't possible here?

This is not within the test we're talking about. This test checks the code that generates the fanout of requests. I could have used various validation code that mod-circulation already has to do preliminary validations so that we don't need to create a fanout list of requests to attempt to place, but I'd rather defer to the existing code to handle this logic than to try to piece them together.

Comment (marcjohnson-kint) [reply src/test/java/api/requests/RequestByInstanceIdResourceTests.java]: I think I still don’t follow. With a known set of items the output should be known?

File: src/main/java/org/folio/circulation/domain/representations/RequestByInstanceIdRequest.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    this.requesterId = requesterId;
+    this.instanceId = instanceId;
+    this.requestExpirationDate = requestExpirationDate;
+    this.pickupServicePointId = pickupServicePointId;
+  }
+
+  public static Result<RequestByInstanceIdRequest> from(JsonObject json) {
+    final DateTime requestDate = getDateTimeProperty(json, REQUEST_DATE);
+
+    if (requestDate == null) {
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/domain/representations/RequestByInstanceIdRequest.java]: Please can tests be added to demonstrate the validation for request date and requester ID

Comment (mtraneis) [reply src/main/java/org/folio/circulation/domain/representations/RequestByInstanceIdRequest.java]: I'll write unit tests for it. The existing API test was merely to check that a failed validation will fail the request gracefully.

**Comment by mreno-EBSCO** [comment]
Sonar reports 3 code smells that we should address before merging.

File: src/main/java/org/folio/circulation/support/ItemRepository.java [hunk 14]
```diff
@@ -41,6 +42,8 @@
   private final boolean fetchMaterialType;
   private final boolean fetchLoanType;
 
+  private static final String items = "items";
```
Comment (marcjohnson-kint) [inline src/main/java/org/folio/circulation/support/ItemRepository.java]:
```suggestion
private static final String ITEMS_COLLECTION_PROPERTY_NAME = "items";
```

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 3 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: What confuses me slightly about that, is the `exceptionally` block does not respond to the client

Comment (mtraneis) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: I'll remove it, then, if it doesn't serve any good purpose.

File: descriptors/ModuleDescriptor-template.json (hunk 2 above)
Comment (mtraneis) [reply descriptors/ModuleDescriptor-template.json]: Added the modulePermissions set

File: descriptors/ModuleDescriptor-template.json [hunk 15]
```diff
@@ -683,6 +683,43 @@
...
+        },
+        {
+          "methods": [
+            "POST"
+          ],
+          "pathPattern": "/circulation/requests/instances",
+          "permissionsRequired": [
+            "circulation.requests.instances.item.post"
+          ],
+          "modulePermissions": [
```
Comment (marcjohnson-kint) [inline descriptors/ModuleDescriptor-template.json]: Does the system fetch the loan type for an item?

If so, we need to add those permissions in to

File: pom.xml [hunk 16]
```diff
@@ -131,6 +131,11 @@
       <version>2.4.0</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
```
Comment (marcjohnson-kint) [inline pom.xml]: What is this used for?

File: descriptors/ModuleDescriptor-template.json (hunk 15 above)
Comment (mtraneis) [reply descriptors/ModuleDescriptor-template.json]: no, it doesn't need loan type

File: pom.xml (hunk 16 above)
Comment (mtraneis) [reply pom.xml]: It was used for a list-ordered map, but after incorporating your changes we don't need it. I'll remove it. I removed it from CIRC-265 but didn't do it here. Good catch

File: src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java (hunk 3 above)
Comment (marcjohnson-kint) [reply src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java]: I think it is a good call to add it.

I have been considering doing it on every top level chain like this.

I think it is more that it needs to write a standard server error response, as well as log
